backend/src/ee/secretRotation/queue.ts

```diff
@@ -0,0 +1,444 @@
+import axios from "axios";
+import Queue, { Job } from "bull";
+import jmespath from "jmespath";
+import { customAlphabet } from "nanoid";
+import { Client as PgClient } from "pg";
+import mysql from "mysql";
+
+import { client, getEncryptionKey, getRootEncryptionKey } from "../../config";
+import { BotService, EventService, TelemetryService } from "../../services";
+import { SecretRotation } from "./models";
+import { rotationTemplates } from "./templates";
+import {
+  ISecretRotationData,
+  ISecretRotationEncData,
+  ISecretRotationProviderTemplate,
+  TAssignOp,
+  TDbProviderClients,
+  TDbProviderFunction,
+  TDirectAssignOp,
+  THttpProviderFunction,
+  TProviderFunction,
+  TProviderFunctionTypes
+} from "./types";
+import {
+  decryptSymmetric128BitHexKeyUTF8,
+  encryptSymmetric128BitHexKeyUTF8
+} from "../../utils/crypto";
+import { ISecret, Secret } from "../../models";
+import { ENCODING_SCHEME_BASE64, ENCODING_SCHEME_UTF8, SECRET_SHARED } from "../../variables";
+import { EESecretService } from "../services";
+import { SecretVersion } from "../models";
+import { eventPushSecrets } from "../../events";
+
+const REGEX = /\${([^}]+)}/g;
+const SLUG_ALPHABETS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+const nanoId = customAlphabet(SLUG_ALPHABETS, 10);
+
+const secretRotationQueue = new Queue("secret-rotation-service", process.env.REDIS_URL as string);
+
+const interpolate = (data: any, getValue: (key: string) => unknown) => {
```

Comment:
I think this can be just one function, can you please simplify so that it calls interpolate on itself? I think this will increase readability for others when adding on to the code base

pseudo code:

```
function interpolate(data: any, variables: ISecretRotationData): any {
  if (!data || typeof data === "number") return data;

  if (typeof data === "string") {
    return data.replace(REGEX, (_, key) => {
      if (key.includes("|")) {
        const [keyword, arg] = key.split("|").map((s: string) => s.trim());
        switch (keyword) {
          case "random":
            return nanoId(parseInt(arg, 10));
          default:
            throw new Error(`Interpolation key not found - ${key}`);
        }
      }

      // Handle references like "inputs.admin_username"
      const [type, keyName] = key.split(".").map((s: string) => s.trim());
      return variables[type as keyof ISecretRotationData][keyName] as string;
    });
  }

  if (typeof data === "object" && Array.isArray(data)) {
    return data.map(item => interpolate(item, variables));
  }

  if (typeof data === "object") {
    ...
  }

  if (typeof data === "object") {
    if ('ref' in data && typeof data.ref === 'string') {
      ...
    } else {
      ...
    }
  }

  return data;
}
```

Reply:
it should not be variables but rather get value because going forward we may introduce more sets like env or some secrets to be pulled etc. 

Then getValue is idea because they don't need to know the original value is